### PR TITLE
Align MLX trainer internals for public UnslothTrainer

### PR DESCRIPTION
**Pull request by Lyxot**
## Summary

Companion PR for the main unsloth MLX trainer API PR: https://github.com/unslothai/unsloth/pull/6462

This updates shared zoo internals used by the new public `UnslothTrainer` route. The goal is to keep MLX behavior close to the CUDA/HF trainer path where notebooks rely on existing `SFTConfig` defaults, while keeping the public routing code in `unsloth` small.

## Companion PRs

- Main unsloth API PR: https://github.com/unslothai/unsloth/pull/6462
- Notebooks PR: https://github.com/unslothai/notebooks/pull/294

## Implementation Details

Shared response masking route:

- Keeps `train_on_responses_only` defined in `unsloth_zoo.dataset_utils` as the shared public helper.
- Detects MLX trainer instances and delegates to `unsloth_zoo.mlx.trainer.train_on_responses_only` for MLX-native label masking.
- Preserves the existing CUDA/HF dataset utility path for non-MLX trainers.
- Handles tokenizer/processor wrappers in the MLX route so notebook processors still resolve to the callable tokenizer needed for response-only masking.

Warmup parity:

- Adds `warmup_ratio` to `MLXTrainingConfig`.
- Resolves warmup ratio after total training steps are known, matching HF `TrainingArguments.get_warmup_steps(total_steps)` behavior.
- Keeps explicit `warmup_steps` authoritative when the caller supplied it, so notebooks that set concrete warmup steps do not get overwritten by a ratio default.
- Uses `ceil(total_steps * warmup_ratio)` when ratio-based warmup is active and steps were not explicit.

Optimizer normalization:

- Centralizes CUDA/Studio optimizer alias handling in `unsloth_zoo.mlx.trainer._normalize_mlx_optimizer_name`.
- Accepts notebook and Studio aliases such as `adamw_8bit`, `paged_adamw_8bit`, `adamw_bnb_8bit`, `paged_adamw_32bit`, `adamw_torch`, `adamw_torch_fused`, `adamw_hf`, and related AdamW aliases by mapping them to MLX `adamw`.
- Keeps native MLX optimizer names (`adamw`, `adam`, `sgd`, `adafactor`, `muon`, `lion`) unchanged.
- Leaves unsupported optimizer names as hard errors with the existing validation path.

Tests touched:

- Updates existing MLX trainer internals coverage for `warmup_ratio`.
- Updates an existing PR684 review-fix test to cover the shared dataset utility delegation to the MLX response-masking route.
- No new zoo test files were added.

## Non-Goals / Deferred Work

- Custom `data_collator` support on MLX remains deferred.
- This PR does not change CUDA/HF dataset masking behavior for non-MLX trainers.
- This PR does not broaden MLX task support beyond the supported notebook SFT surfaces needed by the companion notebook migration.

## Real Gemma3 Notebook Parity Run

I ran a real supported Gemma3 SFT notebook path from `nb/Gemma3_(270M).ipynb` on three variants. The runner follows the notebook cells for model load, LoRA setup, Gemma3 chat template, `Thytu/ChessInstruct` formatting, and `train_on_responses_only`; the only training change is `max_steps=100 -> 10` so the PR can record a short curve. No model weights over 5GB were downloaded.

Run configuration:

- Model: `unsloth/gemma-3-270m-it`
- Dataset: `Thytu/ChessInstruct`, `train[:10000]`
- Filtering after response masking: 665 all-masked rows removed on both CUDA and MLX, leaving 9,335 usable examples
- Sequence length: `2048`
- LoRA: `r=128`, `lora_alpha=128`, target modules `q_proj/k_proj/v_proj/o_proj/gate_proj/up_proj/down_proj`
- Batch: `per_device_train_batch_size=4`, `gradient_accumulation_steps=1`, effective batch size 4
- Scheduler: `warmup_steps=5`, `learning_rate=5e-5`, `lr_scheduler_type="linear"`
- Optimizer argument: `optim="adamw_8bit"`
- Seed: `3407`
- CUDA hardware: RTX 4070 SUPER, Torch `2.11.0+cu128`, Transformers `5.5.0`, TRL `0.24.0`
- MLX hardware: Apple Silicon / MLX, `mlx-lm 0.31.3`, Transformers `5.5.0`, TRL `0.24.0`

Variant commits:

- CUDA main reference: `unslothai/unsloth@45f060899e56d8a8d218529733075814148f6b19` + `unslothai/unsloth-zoo@28063c2a28b148b32390476dc86d2940395416af`, using notebook-main `trl.SFTTrainer` / `SFTConfig`
- CUDA branch: `Lyxot/unsloth@60385efb06d236202a8adc2540a0210935f94511` + `Lyxot/unsloth-zoo@6685266e567511fd2e582ab677834dc0345a62b7`, using branch `UnslothTrainer` / `UnslothTrainingArguments`
- MLX branch retest: `Lyxot/unsloth@1040ccccf0905df2ae4f237fb6953bfad1715827` + `Lyxot/unsloth-zoo@6685266e567511fd2e582ab677834dc0345a62b7`, using branch `UnslothTrainer` / `UnslothTrainingArguments`

Results:

| Step | CUDA main loss | CUDA main grad | CUDA branch loss | CUDA branch grad | MLX branch loss | MLX branch grad |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 4.1118 | 97.9610 | 4.1118 | 97.9610 | 4.0985 | 91.4525 |
| 2 | 3.9207 | 95.0271 | 3.9207 | 95.0271 | 3.9070 | 91.9407 |
| 3 | 3.3180 | 78.2600 | 3.3180 | 78.2600 | 3.3060 | 77.0045 |
| 4 | 1.9192 | 53.1489 | 1.9192 | 53.1489 | 1.8891 | 52.4635 |
| 5 | 1.2203 | 23.5901 | 1.2203 | 23.5901 | 1.1846 | 25.1157 |
| 6 | 0.8083 | 9.4253 | 0.8083 | 9.4253 | 0.8041 | 9.2907 |
| 7 | 0.6018 | 7.3403 | 0.6018 | 7.3403 | 0.5780 | 6.5590 |
| 8 | 0.6660 | 7.1101 | 0.6660 | 7.1101 | 0.6714 | 6.1685 |
| 9 | 0.5907 | 6.9775 | 0.5907 | 6.9775 | 0.5859 | 6.7463 |
| 10 | 0.8521 | 8.8184 | 0.8521 | 8.8184 | 0.8565 | 8.8934 |

Curve summary:

- CUDA branch vs CUDA main: max absolute loss delta `0.0`; max absolute grad-norm delta `0.0`.
- MLX branch vs CUDA main: same curve shape with max absolute loss delta `0.0357`; mean absolute loss delta `0.0147`; final loss `0.8565` vs CUDA `0.8521`; final grad norm `8.8934` vs CUDA `8.8184`.
- MLX retest trained tokens: `481`, matching the previous 30-step parity run's first 10 steps after removing the extra EOS token.
- Average train loss: CUDA main `1.8009`, CUDA branch `1.8009`, MLX branch `1.7881`.

## Validation

Focused validation performed with this zoo head and the paired unsloth head:

- Zoo optimizer alias smoke through `unsloth_zoo.mlx.trainer._normalize_mlx_optimizer_name`:
  - `adamw_8bit -> adamw`
  - `paged_adamw_8bit -> adamw`
  - `adamw_torch_fused -> adamw`
  - native `adamw`, `sgd`, `lion` remain unchanged
- Paired unsloth public trainer API tests:
  - `35 passed, 1 skipped`
- Paired Studio MLX worker config tests:
  - `14 passed`
- `git diff --check` passed in the zoo worktree.
- Gemini review cleanup focused tests after `b0a504da`: `2 passed` for warmup-ratio scheduling and MLX response-masking delegation.
- Codex/Copilot review cleanup focused tests after `6685266e`: `3 passed` for config fields, warmup-ratio scheduling, explicit default warmup preservation, clamping, and MLX response-masking delegation.

The full zoo CI was still running when this description was rewritten; earlier completed zoo jobs on the previous head were green except for the intentionally skipped Mac MLX install job.

### CODE REVIEW

**Comment by Lyxot** [comment]
@codex review

**Comment by Lyxot** [comment]
@codex review

**Comment by Lyxot** [comment]
@codex review

File: unsloth_zoo/mlx/trainer.py [hunk 1]
```diff
@@ -677,10 +695,38 @@ def _ensure_lora_frozen(model):
...
+        except (TypeError, ValueError):
+            return max(0, warmup_steps)
+        if warmup_ratio == 0.0:
+            return max(0, warmup_steps)
+
+        default_warmup_steps = getattr(type(self.args), "warmup_steps", 5)
+        steps_explicit = getattr(
+            self.args,
+            "_unsloth_mlx_warmup_steps_explicit",
+            warmup_steps != default_warmup_steps,
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/trainer.py:896]: Already fixed in `6685266e`. `MLXTrainingConfig.__init__` records whether `warmup_steps` was actually supplied, so `MLXTrainingConfig(warmup_steps=5, warmup_ratio=0.1)` preserves 5. The focused scheduler test covers explicit default preservation and clamping.

**Comment by danielhanchen** [comment]
Reviewed and tested. The MLX internals here are correct and the consumers (public `UnslothTrainer` and the migrated notebooks) work on real Apple Silicon.

## Verified

- Full zoo MLX unit suites pass under the torch simulation shim: `tests/test_mlx_trainer_internals.py` 52 passed, `tests/test_pr684_review_fixes_a.py` 17 passed.
- On a macos-14 runner with real wheels (mlx 0.31.2, full run https://github.com/danielhanchen/unsloth-staging-2/actions/runs/27955271488): `_normalize_mlx_optimizer_name` maps `adamw_8bit` / `paged_adamw_8bit` / `adamw_torch_fused` etc to `adamw`, keeps native names, and raises on unknown; `MLXTrainingConfig.warmup_ratio` plus explicit-warmup tracking behaves as intended; `train_on_responses_only` delegates to the MLX route; and a tiny LoRA SFT trains cleanly with the correct warmup + linear schedule.
- Confirmed the hand-written `MLXTrainingConfig.__init__` does not conflict with `@dataclass` (dataclass does not generate `__init__` when one is defined in the class body), and the class-level field defaults still resolve.

## Minor, non-blocking

1. `train_on_responses_only` probes the MLX trainer with `except ImportError` on every call. On a clean CUDA host this correctly falls back to the HF path. If a half-installed mlx raises something other than `ImportError` (e.g. a Metal init `RuntimeError`) it would propagate into the CUDA path. Consider broadening to `except Exception` and caching the probe result so it does not re-import per call.
2. The optimizer alias mapping is an explicit allow-list, so a future `adamw_*` variant would hard-error. A `startswith("adamw")` fallback would be more forgiving. Note the unsloth Studio worker now delegates to this same function, so the behavior is shared.

Companion PRs unsloth#6462 and notebooks#294 reviewed as well.

**Comment by Lyxot** [comment]
Pushed the latest local MLX trainer updates on top of the current PR branch.

Commits included in this update:
- `fda2984c` `fix: expose MLX response-masked dataset`
- `3a8b8d60` `fix: return trainer-compatible MLX metrics`
- `7d18a945` `fix: cover MLX notebook trainer surfaces`
- `f3a549ff` `fix: support raw MLX notebook inference cells`
- `8a772043` `fix: support MLX VLM generate shim`
- `53e8b4cc` `fix: use bundled gguf for mlx export`

Summary:
- Exposes the MLX response-masked dataset route used by `train_on_responses_only`.
- Aligns MLX trainer metrics with the notebook/Trainer-compatible shape.
- Covers notebook-facing trainer surfaces including standard text training, RAFT-style args, collator acceptance, and `packing=True` fallback behavior.
- Supports raw notebook inference/save cells and the VLM `generate` shim so migrated notebooks can run without user-side MLX edits.
- Fixes MLX GGUF export in uv environments by using the bundled `llama.cpp/gguf-py` path during conversion instead of trying `python -m pip install ...` at runtime.

Validation:
- Text `UnslothTrainer` one-step training passed.
- RAFT-style `TrainingArguments` coercion passed.
- VLM one-step training passed with `UnslothVisionDataCollator`.
- Text and VLM generation/save cells passed with notebook-style CUDA-shaped inputs.
- `python -m pytest tests/test_mlx_save_export_regressions.py -q`: 30 passed.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review
